sycl/test-e2e/Graph/Update/dyn_work_group_memory_usage.cpp

```diff
@@ -0,0 +1,76 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=0 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+// Extra run to check for immediate-command-list in Level Zero
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+
+// Tests updating dynamic_work_group_memory with a new size.
+
+#include "../graph_common.hpp"
+#include <sycl/group_barrier.hpp>
+
+namespace exp_ext = sycl::ext::oneapi::experimental;
+using namespace sycl;
+
+int main() {
+  queue Queue{};
+
+  constexpr int N{1024};
+  constexpr int LocalSize{16};
+
+  using T = int;
+
+  exp_ext::command_graph Graph{Queue.get_context(), Queue.get_device()};
+
+  int *PtrA = malloc_device<T>(N, Queue);
+  std::vector<T> HostDataA(N);
+
+  exp_ext::dynamic_work_group_memory<T[]> DynLocalMem{Graph, LocalSize};
+
+  Queue.memset(PtrA, 0, N * sizeof(T)).wait();
+
+  nd_range<1> NDRange{range{N}, range{LocalSize}};
+
+  auto KernelNode = Graph.add([&](handler &cgh) {
+    cgh.parallel_for(NDRange, [=](nd_item<1> Item) {
+      size_t GlobalID = Item.get_global_id();
+      auto LocalRange = Item.get_local_range(0);
+      auto LocalMem = DynLocalMem.get();
```

Comment:
Could we have a unittest to check the error condition from the `.get()` section of the spec
```
throws an exception with the `errc::invalid` error code if used
in host code.
```

Our unittest are currently disabled due to a CI issue on Windows, so you might need to re-enable them locally to develop this new tests.